### PR TITLE
Upgrade to etcd/bbolt 

### LINKED ISSUE #11: update the boltdb import to bbolt
https://github.com/etcd-io/bbolt

bbolt is supported by coreos.

EDIT:
i see there is a PR for this already !!!

### LINKED ISSUE #21: NewBoltStore failing race condition test with error `converted pointer straddles multiple allocations`.
Hi,

When opening a new file in `NewBoltStore` `-race condition test fails`

Go Version Used: `go version go1.15.6 linux/amd64`

How To reproduce locally.

```
import (
	"fmt"
	"os"
	"path/filepath"
	boltstore "github.com/hashicorp/raft-boltdb"
)
func NewStorage() {
	dir := os.TempDir()
	stableStore, err := boltstore.NewBoltStore(filepath.Join(dir, "test.file"))
	if err != nil {
		panic(err)
	}
	fmt.Println(stableStore)
}
```

Test:

```
func TestStraddles(t *testing.T) {
	t.Parallel()
	NewStorage()
}
```

Run Test with `-race` flag. `go test -v err.go err_test.go  -race`

test will panic.

```
=== RUN   TestStraddles
=== PAUSE TestStraddles
=== CONT  TestStraddles
fatal error: checkptr: converted pointer straddles multiple allocations
goroutine 5 [running]:
runtime.throw(0xc1b46f, 0x3a)
	/usr/lib/go/src/runtime/panic.go:1116 +0x72 fp=0xc000053a18 sp=0xc0000539e8 pc=0x451f52
runtime.checkptrAlignment(0xc0000286d0, 0xbc54a0, 0x1)
	/usr/lib/go/src/runtime/checkptr.go:20 +0xc9 fp=0xc000053a48 sp=0xc000053a18 pc=0x420f29
github.com/boltdb/bolt.(*Bucket).write(0xc000053bb8, 0x0, 0x0, 0x0)
	/home/ankur/go/pkg/mod/github.com/boltdb/bolt@v1.3.1/bucket.go:626 +0x15f fp=0xc000053ab0 sp=0xc000053a48 pc=0x64213f
github.com/boltdb/bolt.(*Bucket).CreateBucket(0xc0002ea018, 0xf63dd8, 0x4, 0x4, 0xc00000c1e0, 0xc0002ea000, 0x0)
	/home/ankur/go/pkg/mod/github.com/boltdb/bolt@v1.3.1/bucket.go:188 +0x32d fp=0xc000053c68 sp=0xc000053ab0 pc=0x63e58d
github.com/boltdb/bolt.(*Bucket).CreateBucketIfNotExists(0xc0002ea018, 0xf63dd8, 0x4, 0x4, 0x0, 0xc000021050, 0xc00000f420)
	/home/ankur/go/pkg/mod/github.com/boltdb/bolt@v1.3.1/bucket.go:206 +0x5e fp=0xc000053cd0 sp=0xc000053c68 pc=0x63ea1e
github.com/boltdb/bolt.(*Tx).CreateBucketIfNotExists(...)
	/home/ankur/go/pkg/mod/github.com/boltdb/bolt@v1.3.1/tx.go:115
github.com/hashicorp/raft-boltdb.(*BoltStore).initialize(0xc00000f420, 0x0, 0x0)
	/home/ankur/go/pkg/mod/github.com/hashicorp/raft-boltdb@v0.0.0-20171010151810-6e5ba93211ea/bolt_store.go:98 +0x128 fp=0xc000053d88 sp=0xc000053cd0 pc=0x8e2b08
github.com/hashicorp/raft-boltdb.New(0xc000024420, 0xe, 0x0, 0xc000024400, 0xe, 0xc000001710, 0xc000001610)
	/home/ankur/go/pkg/mod/github.com/hashicorp/raft-boltdb@v0.0.0-20171010151810-6e5ba93211ea/bolt_store.go:81 +0x1bb fp=0xc000053e28 sp=0xc000053d88 pc=0x8e289b
github.com/hashicorp/raft-boltdb.NewBoltStore(...)
	/home/ankur/go/pkg/mod/github.com/hashicorp/raft-boltdb@v0.0.0-20171010151810-6e5ba93211ea/bolt_store.go:60
github.com/ankur-anand/dkvdb.NewStroage()
	/home/ankur/Documents/github@ankur-anand/go-workspace/dkvdb/err.go:13 +0xf2 fp=0xc000053eb8 sp=0xc000053e28 pc=0xa55832
command-line-arguments_test.TestStraddles(0xc000001680)
	/home/ankur/Documents/github@ankur-anand/go-workspace/dkvdb/err_test.go:11 +0x3e fp=0xc000053ed0 sp=0xc000053eb8 pc=0xa55a7e
testing.tRunner(0xc000001680, 0xc24ea8)
	/usr/lib/go/src/testing/testing.go:1123 +0x203 fp=0xc000053fd0 sp=0xc000053ed0 pc=0x5b0ec3
runtime.goexit()
	/usr/lib/go/src/runtime/asm_amd64.s:1374 +0x1 fp=0xc000053fd8 sp=0xc000053fd0 pc=0x48b421
created by testing.(*T).Run
	/usr/lib/go/src/testing/testing.go:1168 +0x5bc
goroutine 1 [chan receive]:
testing.tRunner.func1(0xc000001500)
	/usr/lib/go/src/testing/testing.go:1088 +0x333
testing.tRunner(0xc000001500, 0xc000377ce0)
	/usr/lib/go/src/testing/testing.go:1127 +0x22b
testing.runTests(0xc00000f400, 0xf4e980, 0x1, 0x1, 0xc0083dfb0d990b3d, 0x8bb384cd56, 0xf98000, 0xc000253308)
	/usr/lib/go/src/testing/testing.go:1437 +0x613
testing.(*M).Run(0xc000200180, 0x0)
	/usr/lib/go/src/testing/testing.go:1345 +0x3b4
main.main()
	_testmain.go:45 +0x237
goroutine 6 [chan receive]:
testing.runTests.func1.1(0xc000001500)
	/usr/lib/go/src/testing/testing.go:1444 +0x65
created by testing.runTests.func1
	/usr/lib/go/src/testing/testing.go:1444 +0xdb
FAIL	command-line-arguments	0.023s
FAIL
```

### PR DESCRIPTION
**Pull request by s-christoff**
This closes #11. 
Fixes #21
Thank you to #12, #19, #13.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -3,8 +3,8 @@ module github.com/hashicorp/raft-boltdb
 go 1.12
 
 require (
-	github.com/boltdb/bolt v1.3.1
+	// github.com/boltdb/bolt v1.3.1
```
Comment (rboyer) [inline go.mod:6]: Did you intentionally leave this behind?

Comment (s-christoff) [reply go.mod:6]: whoops

Comment (s-christoff) [reply go.mod]: fixed!

**Comment by s-christoff** [comment]
I reverted the initial changes and made the necessary changes to do it [this way](https://blog.golang.org/using-go-modules)

**Review by rboyer — APPROVED** [review]
LGTM but maybe also get someone else raft-y to :+1: it too

File: v2/README.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+raft-boltdb/v2
+===========
+
+This implementation uses the maintained version of BoltDB, [BBolt](https://github.com/etcd-io/bbolt). This is the primary version of `raft-boltdb` and should be used whenever possible.
```
Comment (banks) [inline v2/README.md:4]: Worth mentioning why it's a v2? I.e. there is no breaking API change to the library it's just the potential for disk format incompatibilities that means we are being conservative and making it a separate import path.

The separate import path is important vs, just a minor version bump because it allows anyone who needs to to import both versions to perform a safe in-place upgrade of old files read with the old version and written back out with the new one.

**Review by banks — COMMENTED** [review]
I think this looks good Sarah. One downside of the v2 thing is that It's hard to tell in this PR whether you actually changed any code or tests here or just copied and switched out the dependency.

Can you confirm if there were any code changes needed so we can look a little more closely there?
Could y

File: v2/bolt_store.go [hunk 3]
```diff
@@ -0,0 +1,268 @@
+package raftboltdb
+
+import (
+	"errors"
+
+	"go.etcd.io/bbolt"
```
Comment (s-christoff) [inline v2/bolt_store.go:6]: I'mma just point out all the changes to bbolt so it's a bit easier to read, so here's 1 (there are not many)

File: v2/bolt_store.go [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+	// An error indicating a given key does not exist
+	ErrKeyNotFound = errors.New("not found")
+)
+
+// BoltStore provides access to BoltDB for Raft to store and retrieve
+// log entries. It also provides key/value storage, and can be used as
+// a LogStore and StableStore.
+type BoltStore struct {
+	// conn is the underlying handle to the db.
+	conn *bbolt.DB
```
Comment (s-christoff) [inline v2/bolt_store.go:30]: bbolt change

**Comment by s-christoff** [comment]
Here's a quick little guide on the changes in the v2 folder - 
Here's the [changed import](https://github.com/hashicorp/raft-boltdb/pull/22/files#diff-cd3e2fa22ef4a74a78b671e33b9450c71b147125fde4996e3bd395d9f3da14e5R6) , vs [v1](https://github.com/hashicorp/raft-boltdb/blob/master/bolt_store.go#L6)
Update the [BoltStore struct](https://github.com/hashicorp/raft-boltdb/pull/22/files#diff-cd3e2fa22ef4a74a78b671e33b9450c71b147125fde4996e3bd395d9f3da14e5R30), vs [v1](https://github.com/hashicorp/raft-boltdb/blob/master/bolt_store.go#L30)
Also update the [options](https://github.com/hashicorp/raft-boltdb/pull/22/files#diff-cd3e2fa22ef4a74a78b671e33b9450c71b147125fde4996e3bd395d9f3da14e5R43), vs [v1](https://github.com/hashicorp/raft-boltdb/blob/master/bolt_store.go#L43)
The only place it's called afaik is in [New()](https://github.com/hashicorp/raft-boltdb/pull/22/files#diff-cd3e2fa22ef4a74a78b671e33b9450c71b147125fde4996e3bd395d9f3da14e5R66), vs [v1](https://github.com/hashicorp/raft-boltdb/blob/master/bolt_store.go#L66)

Then similarly in the test, [changed import](https://github.com/hashicorp/raft-boltdb/pull/22/files#diff-cd3e2fa22ef4a74a78b671e33b9450c71b147125fde4996e3bd395d9f3da14e5R6) - and throughout the test `bbolt.Options` is called. 

tl;dr - Because of the `BoltStore{}` struct and `Options{}` exist there are in minimal changes. My [initial commit](https://github.com/hashicorp/raft-boltdb/pull/22/commits/22de16c8e211a122225ee5868ff64b01bf837c3f) can be put next to this commit for reference as well - which may help with clarity.

**Review by banks — APPROVED** [review]
Awesome, thanks for the guide @s-christoff that looks very cool.
